feat(scripts) upload bundle script

loweisz opened the pull request:
Adds the upload script to `app-scripts` in interactive and in non-interactive (`--ci`) mode.

shikaan commented on packages/@contentful--app-scripts/lib/upload/index.js:
```diff
@@ -0,0 +1,15 @@
+    const settings = await buildAppUploadSettings();
+    await createAppBundle(settings);
+  },
+  async nonInteractive(program, options) {
```
Why do we need `program` here?

loweisz replied:
I used it in a previous implementation, forgot to remove it

shikaan commented on packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js:
```diff
@@ -0,0 +1,60 @@
+  Run ${chalk.dim('contentful-app-scripts upload --help')} to see all required arguments
+        `
+      );
+      throw new Error('Invalid Arguments');
```
nit: this might be a `TypeError`

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
+    const filePaths = await new Promise((res) => {
+      glob(path + '/**/*', { nodir: true }, (err, paths) => {
+        res(paths);
+      });
+    });
```
You can use `util.promisify` here to stick with the async/await style we have elsewhere

loweisz replied:
it's not used anymore anyway :D

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+
+async function convertPathToFileData(filePath) {
+  return new Promise((res) => {
+    fs.readFile(filePath, 'utf8', (err, data) => {
+      res([filePath, data]);
+    });
+  });
```
You can use `fs/promises` instead of wrapping. Although, being a script, also sync API would do

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
```
You can use [`adm-zip`](https://github.com/cthackers/adm-zip) here: it's MIT licensed and it allows adding a full folder by its path (so no need to do `fs` manipulation)

loweisz replied:
that was a really good tip 😬  Could remove half of the code haha

shikaan commented on packages/@contentful--app-scripts/lib/upload/definition-api.js:
```diff
@@ -0,0 +1,66 @@
+const inquirer = require('inquirer');
+const ora = require('ora');
+
+async function fetchDefinitions(client, orgId) {
```
Why do we duplicate this?

shikaan replied:
Also, for the rest of the file, see comments on `organization-api`

shikaan commented on packages/@contentful--app-scripts/lib/upload/organization-api.js:
```diff
@@ -0,0 +1,72 @@
+const chalk = require('chalk');
+const ora = require('ora');
+
+async function fetchOrganizations(client) {
```
Also using the error wrapper I was suggesting below and the plain client in CMA, this can become a couple of lines of code. You might not need this utility in the first place

loweisz replied:
I actually already did that in the second PR about activation. But I moved it now here too

shikaan commented on packages/@contentful--app-scripts/lib/upload/organization-api.js:
```diff
@@ -0,0 +1,72 @@
+  }
+}
+
+async function selectOrganization(client) {
+  const orgSpinner = ora('Fetching your organizations').start();
+  let organizations = null;
+  try {
+    organizations = await fetchOrganizations(client);
+  } finally {
+    orgSpinner.stop();
+  }
+
+  if (!organizations) {
+    return null;
+  }
+  const { organizationId } = await inquirer.prompt([
+    {
+      name: 'organizationId',
+      message: 'Select an organization for your app:',
+      type: 'list',
+      choices: organizations,
+    },
+  ]);
+  return organizations.find((org) => org.value === organizationId);
+}
```
Selecting from a list is something we do quite often. I would extract a utilty for it and reuse it for definitions and organizaions

loweisz replied:
nice idea I abstracted it in utils 👍

shikaan commented on packages/@contentful--app-scripts/lib/upload/index.js:
```diff
@@ -0,0 +1,15 @@
+    const settings = await buildAppUploadSettings();
+    await createAppBundle(settings);
+  },
+  async nonInteractive(program, options) {
```
Why `program` here?

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+const { showCreationError } = require('../utils');
+
+async function convertPathToFileData(filePath) {
+  return new Promise((res) => {
```
You can use `fs/promises` instead of wrapping. Although, being a script, also sync API would do

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
```
You can use [`adm-zip`](https://github.com/cthackers/adm-zip) here: it's MIT licensed and it allows adding a full folder by its path (so no need to do `fs` manipulation)

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-zip-from-directory.js:
```diff
@@ -0,0 +1,61 @@
+async function createZipFileFromDirectory(path) {
+  try {
+    let zip = new JSZip();
+    const filePaths = await new Promise((res) => {
```
You can use `util.promisify` here to stick with the async/await style we have elsewhere

shikaan reviewed (commented):
It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)

One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?

loweisz commented on packages/@contentful--app-scripts/lib/upload/organization-api.js:
```diff
@@ -0,0 +1,72 @@
+)} Could not fetch your organizations. Make sure you provided a valid access token.
+
+${err.message}
+    `);
```
Good idea 👍

loweisz commented:
> It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> 
> One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?

@shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬

But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?

shikaan commented:
> > It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> > One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?
> 
> @shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬
> 
> But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?

I was more thinking about we don't ask for the things we already know. So, if something is passed (even in interactive mode), we don't request further details on that.

TL;DR: we keep the same interface for interactive and non-interactive commands, but in the interactive case the arguments are all optional. WDYT?

loweisz commented:
> > > It's a good starting point. There are a couple fo things that can be streamlined and then we are ready to ship this :)
> > > One question: what is it going to happen when we use the interactive mode _and_ we pass the parameters? Will it throw or will it just ask for them?
> > 
> > 
> > @shikaan It actually will just ask for them and ignore the ones passed. I tried to make them conditional that when you pass `--ci` you will get a better notification that you have to pass them in that case. But somehow I couldn't find a good way 😬
> > But I think we could throw when passed arguments without `--ci`, do you think that would make more sense than just ignoring it?
> 
> I was more thinking about we don't ask for the things we already know. So, if something is passed (even in interactive mode), we don't request further details on that.
> 
> TL;DR: we keep the same interface for interactive and non-interactive commands, but in the interactive case the arguments are all optional. WDYT?

Ahh I see ! That's a good idea, will look into it

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:
```diff
@@ -0,0 +1,72 @@
+async function createAppBundleFromUpload(settings, appUploadId) {
+  const clientSpinner = ora('Verifing your upload').start();
+  const client = createClient({ accessToken: settings.accessToken });
+  const org = await client.getOrganization(settings.organisation.value);
```
nit: organi**z**ation for consistency

shikaan reviewed (commented):
Looks great 🎉  We only need to cover it with some tests and we're good to go 🚢

andipaetzold commented on packages/@contentful--app-scripts/lib/upload/get-upload-settings-args.js:
```diff
@@ -0,0 +1,60 @@
+      console.log(
+        `
+  ${chalk.red('Invalid Arguments:')} the argument ${chalk.cyan(argument)} was not defined
+  Run ${chalk.dim('contentful-app-scripts upload --help')} to see all required arguments
```
That command only works when the `contentful-app-scripts` are installed globally. If there is a chance that this error is displayed when using the CCA output, we should provide an alternative call here. Even `npx contentful-app-scripts --help` might do it.

Also: This should be `@contentful/app-scripts`

loweisz replied:
True ! I was planning to change that with the real command, changed it now

loweisz replied:
True ! I was planning to change that with the real command, changed it now

andipaetzold commented on packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:
```diff
@@ -0,0 +1,72 @@
+
+    ${chalk.bold('or by simply running the cli command:')}
+
+      ${chalk.underline(`contentful-app-scripts activate`)}
```
Same issue with the app scripts package here.

andipaetzold commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
+    .description('Upload your build folder and create an app bundle')
+    .option('--bundle-dir [directory]', 'The directory of your build folder')
+    .option('--organization-id [orgId]', 'The id of your organisation')
+    .option('--definition-id  [defId]', 'The id of your apps definition')
```
```suggestion
    .option('--definition-id [defId]', 'The id of your apps definition')
```

andipaetzold commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
 
-function getCommand(program, command) {
+
+function runCommand(command, options) {
```
This must be async and await for `nonInteractive()` and `interactive()`

andipaetzold commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
-    const command = getCommand(program, createAppDefinition);
+    .command('create-app-definition')
+    .description('Create a new AppDefinition for an App')
+    .action((options) => {
+      runCommand(createAppDefinition, options);
+    })
```
Should be an async function with await

andipaetzold commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
+    .option('--definition-id  [defId]', 'The id of your apps definition')
+    .option('--token [accessToken]', 'Your content management access token')
+    .option('--comment [comment]', 'Optional comment for the created bundle')
+    .action((options) => {
+      runCommand(upload, options);
+    })
```
Should be an async function with await

andipaetzold reviewed (changes requested):
You alternate between using `module.exports = ` and `exports.xxx`. Any reason for that?
I'd make this consistent and always use `module.exports`

Also, whether a log line has a trailing period `.` is inconsistent.

andipaetzold commented on packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:
```diff
@@ -9,8 +9,8 @@ async function createAppBundleFromUpload(settings, appUploadId) {
   const client = createClient({ accessToken: settings.accessToken });
-  const org = await client.getOrganization(settings.organisation.value);
-  const appDefinition = await org.getAppDefinition(settings.definition.value);
+  const organization = await client.getOrganization(settings.organisation.value);
```
Manuel was talking about `organisation` with an `s` here.

andipaetzold replied:
The other way round. We use American English in contentful. So: `organization`.
Sorry for being unspecific.

loweisz replied:
Ahhh I see, okay I changed it now everywhere to `organization`

loweisz commented:
> You alternate between using `module.exports = ` and `exports.xxx`. Any reason for that?
> I'd make this consistent and always use `module.exports`
> 
> Also, whether a log line has a trailing period `.` is inconsistent.

Not really `exports` was used in the code base before. I changed it to `module.exports` now everywhere and also added a `.` to all logs 👍

shikaan commented on packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:
```diff
@@ -0,0 +1,77 @@
+  let appUpload = null;
+  try {
+    appUpload = await createAppUpload(settings);
+    console.log('UPLOAD', appUpload);
```
Is this leftover? 🤔

shikaan commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
-    const command = getCommand(program, createAppDefinition);
+    .command('create-app-definition')
+    .description('Create a new AppDefinition for an App')
+    .action(async (options) => {
+      await runCommand(createAppDefinition, options);
+    })
```
super-nit

```suggestion
    .action((options) => runCommand(createAppDefinition, options))
```

shikaan commented on packages/@contentful--app-scripts/bin/app-scripts:
```diff
@@ -3,32 +3,41 @@
-function getCommand(program, command) {
+
+async function runCommand(command, options) {
   const {ci} = program.opts()
-  return ci ? command.nonInteractive : command.interactive;
+  return ci ? await command.nonInteractive(options) : await command.interactive(options);
 }
```
nit: I think this was good as it was: you're returning whatever the command returns, no need to `await`

andipaetzold replied:
I always prefer adding await on every level. That way we get a proper stack trace. The outcome when running the code will be the same.

loweisz replied:
I think in the end it doesn't really matter when in both cases `runCommand` will return a promise, or am I missing something ? @andipaetzold what was your intention to use await here?

shikaan reviewed (approved):
Great job 🎉 

There is only one - possible - leftover and a couple of `nit`s (mostly for code hygiene, no blockers)

andipaetzold commented on packages/@contentful--app-scripts/lib/create-app-definition/create-app-definition.js:
```diff
@@ -36,18 +36,24 @@ function assertValidArguments(accessToken, appDefinitionSettings) {
+        appDefinitionSettings,
+        null,
+        2
+      )}``Example: ${JSON.stringify(
```
Does this syntax with double ` work?

loweisz replied:
I actually haven't changed something here it looks like it came from prettier when I changed something else in the file, so it probably will work when it worked before :D

andipaetzold replied:
Well it was a double ` before, so it didn't change. But e.g. `console.log(`test``test`)` throws an error for me as that's the syntax for template literals: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

loweisz replied:
Yes you are right, I think there was a `,` lost as `throwValidationException` is also accepts three arguments. 
I added it now that the example is the `details` in the function

andipaetzold commented on packages/@contentful--app-scripts/lib/upload/create-app-bundle.js:
```diff
@@ -64,9 +66,12 @@ async function createAppBundle(settings) {
     ${chalk.bold('or by simply running the cli command:')}
 
-      ${chalk.underline(`contentful-app-scripts activate`)}
+      ${chalk.underline(`'npx @contentful/app-scripts activate'`)}
```
```suggestion
      ${chalk.underline('npx @contentful/app-scripts activate')}
```
